modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -395,7 +399,12 @@ public CompletableFuture<Void> finish(
                         enlistedGroups,
                         txId,
                         finishingStateMeta.txFinishFuture()
-                ));
+                )
+        ).thenAccept(unused -> {
+            if (localNodeId.equals(finishingStateMeta.txCoordinatorId())) {
```

Comment:
I believe that it'll work. However in case of using txId as in inner counter identifier instead of beginTs you may just skip this check. Decrementing the counter wihtin recovery on non coordinator would just have no effect because there won't be any transactions with given id. Currently while using beginTs it won't work, because beginTs is unique only within node. Up to you.

Reply:
I made changes about the independence of the decrement, I think this should be enough.